dev-docs/RFCs/v7.3/program-manager-rfc.md

```diff
@@ -0,0 +1,125 @@
+
+
+## Background
+
+Creation and management of programs is currently handled by several systems in luma.gl:
+- Modules are simply JavaScript objects that can be created anywhere by the application.
+- Injections and shader hooks are managed as library-wide global state managed by functions `setShaderHook` and `setModuleInjection`.
+- Assembling base shader sources, modules, defines and injections into a final shader source is handled by `assembleShaders`.
+- Caching and re-use of programs is handled by `ShaderCache`, which is currently limited in that it scopes caches to `BaseModel` objects meaning that `BaseModel`s cannot share programs (this is due to an easily-remedied limitation in the `BaseModel` class that it doesn't set uniforms per-draw).
+
+Issues with the current system include:
+- These systems are currently only consolidated around the `BaseModel` class, which treats programs as immutable. This means that modifying programs and taking advantage of these system involves re-constructing the entire `BaseModel`. This complicates interaction with programs as they and parameters that modify their source must generally be treated as immutable and separated out from mutable properties (e.g. the [data filter layer extension](https://github.com/uber/deck.gl/blob/6113d2c8984c406e9df59c16f19630a18f36c42d/modules/extensions/src/data-filter/data-filter.js#L42) in deck.gl has to handle `filterSize`  and `softMargin` separately from other props because they modify the shader code).
+- Shader hooks and module injections are global state meaning that an application that imports luma.gl must use the same definitions globally (a case where one might not want this is when rendering to separate canvases on the same page).
+- Caching being limited per-model presents serious performance concerns as program compilation is expensive, and program switching is among the more expensive GL state changes (this would be a concern, for example, in deck.gl when rendering multiple instances of the same layer).
+- With out a centralized system for managing programs, it will be difficult to take advantage of optimizations that require batching program operations, such `KHR_parallel_shader_compile`.
+
+
+## Overview
+
+The proposed program manager would take the caching mechanisms provide by `ShaderCache` and extend it in the following ways:
+- Remove the per-model scoping.
+- Manage shader sources, modules, injections and shader hooks.
+- Provide a simple API for building a program from shader sources, modules and defines, returning a cached version if available.
+
+The `BaseModel` class would be modified to:
+- Support attachment of a program manager from which it would acquire programs.
+- Support sharing programs with other `BaseModel`s.
+- Support switching programs.
+
+## Implementation
+
+A `ProgramManager` that supports the following methods:
+- `addVertexShader(id, source)`: register vertex shader source code.
```

Comment:
Nit: This seems like a lot of methods to define and document for an API that will "never" be called by the user (I'm assuming 99% use will be inside the BaseModel class). 

Maybe just something like:
```js
new ProgramManager(gl)

programManager.register({
  vertexShaders: {id: source, ...},
  fragmentShaders: {id: source ...};
  modules: [],
  shaderHooks: []
));
```

Reply:
I don't think so. I expect usage to be more like the following:
- deck registers modules and shader hooks
- layers register their shaders
- `BaseModel`  `get`s and `release`s programs.